###################################################################################################################################
#
# Copyright 2017 IRD and Grenoble-Alpes University
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/> or
# write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.
#
# You should have received a copy of the CeCILL-C license with this program.
# If not see <http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.txt>
#
# Intellectual property belongs to IRD and Grenoble-Alpes University
#
# Written by Philippe Cubry, Yves Vigouroux and Olivier Francois, modified by Nora Scarcelli
#
###################################################################################################################################

######################################################
# R script to compute set of frequencies and summary statistics on observed datasets for Splatche analysis  
######################################################

library("data.table")

# Function to compute minimum allele frequency for diploid data
maf4 = function(x){ 
  x = x[x != 9]
  f = sum(x) ; (min(f , 2*length(x) - f ))}

# Function to compute missing value frequency
miss.freq = function(x){ 
  n = length(x)
  x = x[x == 9]
  f = length(x)/n}


# Function for binning SFS
#### sapply(1:xx) with xx = number of individuals
#### c(y[1], ... ) with approriate number of SFS bins
histo.bin = function(x){
  n = length(x)
  y = sapply(1:80, FUN = function(i) sum(x == i) )
  c(y[1], y[2], sum(y[3:4]), sum(y[5:11]), sum(y[12:21]), sum(y[22:35]), sum(y[36:55]), sum(y[56:80]) )/n
}


#### sapply(1:xx) with xx = number of groups for rare variant calculation
z.group = function(z){
  sapply(1:11, FUN = function(i) sum(z[group == i]) )
}

#### sapply(1:xx) with xx = number of groups for rare variant calculation
z.groupm = function(z){
  sapply(1:11, FUN = function(i) mean(z[group == i]) )
}


#Load list of genotypes with geographic coordinates
samples <-  read.table("./coordinate.txt",header=TRUE,na.strings = "") # File with 6 columns: Sample_Code; Species; Country; Sample_Name; Latitude; Longitude

# Load datasets
genotype <- fread("./vcf.012") # File obtained using vcftools --vcf in.vcf --012 --out out.vcf
genotype=genotype[,-1]

row.names(genotype) <- samples$Sample_Code

# Load group definition for rare variant analysis
group = (read.table("./groups_obs.txt", na.strings="-9", header=TRUE)) # File generated by Group.R
group = as.numeric(as.vector(group[,2]))

# Data transformation
## Tranform df into matrix for faster replacements
genotypes.mat <- as.matrix(genotype) 

genotypes.mat[genotypes.mat == -1] = rep(9, sum(genotypes.mat == -1))

## Compute SFS on cultivated african ##
spect = apply(genotypes.mat, MARGIN = 2, FUN = maf4   )
plot(table(spect),xlab="",ylab="",main = "folded SFS")
# Filtering monomorphic loci
spect.poly <- spect[which(spect!=0)]
plot(table(spect.poly),xlab="",ylab="",main = "folded SFS without monomorphic alleles",las=2)

write.table(spect,"./SFS_obs.txt")

## Compute summary statistics
stat.obs <- NULL

lst1 = which(spect == 1)
geno1 = genotypes.mat[,lst1]

ind1 = apply(geno1, MARGIN = 2, FUN = function(x) {
  y = x[x != 9]
  i1 = which(x != 9)[1]
  i = which( (x != y[1]) & (x != 9) )
  if (length(i) > 1) i1 else i 
})

table(ind1) 
z = sapply(1:nrow(genotypes.mat),
           FUN = function(x) sum(ind1 == x) )

#### c(paste("SFS",seq(1,x,1), sep="") with x= number of SFS bins
#### paste("RareVariants",seq(1,z,1), sep="") with z = number of groups for rare variant calculation
stat.obs = as.data.frame(t(c(histo.bin(spect.poly)/sum(histo.bin(spect.poly)), z.groupm(z)/sum(z.groupm(z)))))
colnames(stat.obs) <- c(paste("SFS",seq(1,8,1), sep=""),paste("RareVariants",seq(1,11,1), sep=""))

write.table(stat.obs,"./sum_stats_obs.txt")
